pkg/sentry/fsimpl/devpts/line_discipline.go

```diff
@@ -450,6 +451,94 @@ func (*inputQueueTransformer) transform(l *lineDiscipline, q *queue, buf []byte)
 			l.terminal.replicaKTTY.SignalForegroundProcessGroup(kernel.SignalInfoPriv(linux.SIGTSTP))
 		case l.termios.ControlCharacters[linux.VQUIT]: // ctrl-\
 			l.terminal.replicaKTTY.SignalForegroundProcessGroup(kernel.SignalInfoPriv(linux.SIGQUIT))
+
+		// In canonical mode, some characters need to be handled specially; for example, backspace.
+		// This roughly aligns with n_tty.c:n_tty_receive_char_canon and n_tty.c:eraser
+		// cBytes[0] == ControlCharacters[linux.VKILL] is also handled by n_tty.c:eraser, but this isn't implemented
+		case l.termios.ControlCharacters[linux.VWERASE]:
+			if !l.termios.LEnabled(linux.IEXTEN) {
+				break
+			}
+			fallthrough
+		case l.termios.ControlCharacters[linux.VERASE]:
+			if !l.termios.LEnabled(linux.ICANON) {
+				break
+			}
+
+			c := cBytes[0]
+			killType := linux.VERASE
+			if c == l.termios.ControlCharacters[linux.VWERASE] {
+				killType = linux.VWERASE
+			}
+			seenAlphanumeric := false
+			for len(q.readBuf) > 0 {
+				// Erase a character. If IUTF8 is enabled, erase an entire multibyte unicode character.
+				var toErase byte
+				cnt := 0
+				isContinuationByte := false
+				for ; cnt < len(q.readBuf) && !isContinuationByte; cnt++ {
```

Comment:
I _think_ we want this to loop when isContinuationByte is true? That is, it should keep consuming bytes as long as the current byte is a continuation. Something like:

```go
var isContinuationByte bool
for isContinuationByte = true; cnt < len(q.readBuf) && isContinuationByte; cnt++ {
  toErase = q.readBuf[len(q.readBuf)-cnt-1]
  isContinuationByte = l.termios.IEnabled(linux.IUTF8) && (toErase&0xc0) == 0x80
}
```

Or maybe I'm misunderstanding. Are the tests you wrote passing?

Reply:
Good catch, thank you! I think you're correct. I fixed this (along with another bug) and now the tests pass; I must have not run the tests correctly last night.